### ISSUE TITLE
Operators can more easily see when routes and endpoints are added or removed from a given gorouter's routing table in the gorouter logs

### ISSUE THREAD

**Issue by mcwumbly** [cloudfoundry/routing-release#172]
# Summary

Our logs have very limited information about the state of the routing table as we only log when routes are _unregistered_ not when they are _registered_. 

In addition, the unregister messages themselves are noisier than need be because we log any time we receive an "unregister-route" message, and these messages are repeated a number of times by the route-emitter make sure they are received.

Rather than logging what _messages_ we receive, it would be nice if instead we had the following `Info` level logs if and only if those messages result in an change to the routing table:

- "route-added"
- "route-removed"
- "endpoint-added"
- "endpoint-removed"

# Expected Results

**Given** A new _route_ is added
**When** a given gorouter adds the route to its routing table
**Then** I see exactly one log message in gorouter.stdout.log indicating the route was added.

**Given** A _route_ is removed or pruned
**When** a given gorouter removes the route from its routing table
**Then** I see exactly one log message in gorouter.stdout.log indicating the route was added.

**Given** An _endpoint_ is added for a route (e.g. when an app is scaled up from N to N+1 instances)
**When** a given gorouter adds the endpoint to its routing table
**Then** I see exactly one log message in gorouter.stdout.log for each endpoint added.

**Given** An _endpoint_ is removed or pruned for a route (e.g. when an app is scaled down from N to N-1 instances)
**When** a given gorouter removes the endpoint from its routing table
**Then** I see exactly one log message in gorouter.stdout.log for each endpoint removed.

# Context

An ongoing source of support burden circles around questions of "stale routes" which are often suspected as a possible cause of folks 'issues when getting errors in their requests to apps, and our documentation for troubleshooting error responses has an entire section devoted to [diagnosing stale routes](https://docs.cloudfoundry.org/adminguide/troubleshooting-router-error-responses.html#diagnose-stale-routes).

The gorouter routing table is eventually consistent (and therefore _often_ inconsistent). The docs cover how to check the routing table at any moment using the `/var/vcap/jobs/gorouter/bin/retreive-local-routes` script that gets put on the router VM. But that only tells you what routes the router has in its table _now_. When handling support issues, we usually start with _logs_.

Our logs have very limited information about the state of the routing table as we only log when routes are _unregistered_ not when they are _registered_. More specifically we log any time we receive an "unregister-route" message.

"register-route" messages and "unregister-route" messages are both sent repeatedly to make sure they are received, so logging them all is noisy. While still noisy, [we do log "unregister-route" messages](https://github.com/cloudfoundry/gorouter/blob/611153b9a6df30752f84618468fa3e88938d8b29/mbus/subscriber.go#L211) because they stop after some time. But "register-route" messages are repeated indefinitely, so we don't log them at all.

Rather than logging what messages we receive at Info level, it would be nice if instead we logged at Info level any time a register-route message results in a route or endpoint actually being added to or removed from the routing table.

**Comment by mcwumbly** [comment]
I've spiked this out and here are the results so far:

branch: https://github.com/cloudfoundry/gorouter/tree/log-route-table-changes
commit: https://github.com/cloudfoundry/gorouter/commit/56c787ded76bdae88b2cc972ec303731c0bddf9a

The log messages below are the _only_ log messages I see now... no more repeated "unregister-route" or "start-pruning" messages...

I am getting a couple duplicate messages when scaling apps up and down, but only single repeats... haven't dug into what's going on there yet.

# Add a new route:

```
$ cf map-route foo springgreen.cf-app.com --hostname a
Mapping route a.springgreen.cf-app.com to app foo in org o / space s as admin...
OK
```

`gorouter.stdout.log`:
```
{"log_level":1,"timestamp":"2020-08-27T22:59:43.462087363Z","message":"route-registered","source":"vcap.gorouter.registry","data":{"uri":"a.springgreen.cf-app.com"}}
{"log_level":1,"timestamp":"2020-08-27T22:59:43.462279999Z","message":"endpoint-registered","source":"vcap.gorouter.registry","data":{"uri":"a.springgreen.cf-app.com","backend":"10.0.1.11:61002","modification_tag":{"guid":"","index":0},"isolation_segment":"-","isTLS":true}}
```

# Scale app from 1 to 2 instances 

```
$ cf scale foo -i 2
Scaling app foo in org o / space s as admin...

Instances starting...

Showing current scale of app foo in org o / space s as admin...

name:              foo
requested state:   started
routes:            foo.springgreen.cf-app.com, a.springgreen.cf-app.com
last uploaded:     Thu 27 Aug 15:27:34 PDT 2020
stack:             cflinuxfs3
buildpacks:        go

type:           web
sidecars:
instances:      1/2
memory usage:   32M
     state      since                  cpu    memory        disk          details
#0   running    2020-08-27T22:59:17Z   0.3%   8.1M of 32M   9.5M of 32M
#1   starting   2020-08-27T22:59:56Z   0.0%   0 of 32M      0 of 32M
```

`gorouter.stdout.log`:
```
{"log_level":1,"timestamp":"2020-08-27T22:59:59.350998043Z","message":"endpoint-registered","source":"vcap.gorouter.registry","data":{"uri":"a.springgreen.cf-app.com","backend":"10.0.1.11:61006","modification_tag":{"guid":"","index":0},"isolation_segment":"-","isTLS":true}}
{"log_level":1,"timestamp":"2020-08-27T22:59:59.351131999Z","message":"endpoint-registered","source":"vcap.gorouter.registry","data":{"uri":"foo.springgreen.cf-app.com","backend":"10.0.1.11:61006","modification_tag":{"guid":"","index":0},"isolation_segment":"-","isTLS":true}}
```

# Scale app from 2 to 1 instances 

```
$ cf scale foo -i 1
Scaling app foo in org o / space s as admin...

Instances starting...

Showing current scale of app foo in org o / space s as admin...

name:              foo
requested state:   started
routes:            foo.springgreen.cf-app.com, a.springgreen.cf-app.com
last uploaded:     Thu 27 Aug 15:27:34 PDT 2020
stack:             cflinuxfs3
buildpacks:        go

type:           web
sidecars:
instances:      1/1
memory usage:   32M
     state     since                  cpu    memory        disk          details
#0   running   2020-08-27T22:59:17Z   0.3%   8.1M of 32M   9.5M of 32M
```

`gorouter.stdout.log`:
```
{"log_level":1,"timestamp":"2020-08-27T23:00:27.122616625Z","message":"endpoint-unregistered","source":"vcap.gorouter.registry","data":{"uri":"a.springgreen.cf-app.com","backend":"10.0.1.11:61006","modification_tag":{"guid":"","index":0},"isolation_segment":"-","isTLS":true}}
{"log_level":1,"timestamp":"2020-08-27T23:00:27.123043785Z","message":"endpoint-unregistered","source":"vcap.gorouter.registry","data":{"uri":"foo.springgreen.cf-app.com","backend":"10.0.1.11:61006","modification_tag":{"guid":"","index":0},"isolation_segment":"-","isTLS":true}}
```

# Remove a route

```
$ cf unmap-route foo springgreen.cf-app.com --hostname a
Removing route a.springgreen.cf-app.com from app foo in org o / space s as admin...
OK
```

`gorouter.stdout.log`:
```
{"log_level":1,"timestamp":"2020-08-27T23:00:46.702876112Z","message":"endpoint-unregistered","source":"vcap.gorouter.registry","data":{"uri":"a.springgreen.cf-app.com","backend":"10.0.1.11:61002","modification_tag":{"guid":"","index":0},"isolation_segment":"-","isTLS":true}}
{"log_level":1,"timestamp":"2020-08-27T23:00:46.703133349Z","message":"route-unregistered","source":"vcap.gorouter.registry","data":{"uri":"a.springgreen.cf-app.com"}}
```

**Comment by mcwumbly** [comment]
> I am getting a couple duplicate messages when scaling apps up and down, but only single repeats... haven't dug into what's going on there yet.

Oh, this is because I had two routes mapped to the app :)

**Comment by jrussett** [comment]
Hi 👋 

This functionality is now available in [Routing Release `0.207.0`](https://github.com/cloudfoundry/routing-release/releases/tag/0.207.0).

Thanks,
@jrussett
